From 91ace97f38270369f43fb15a3a19195ac11915d5 Mon Sep 17 00:00:00 2001
From: Aloka Dixit <quic_alokad@quicinc.com>
Date: Fri, 23 Oct 2020 17:02:06 -0700
Subject: [PATCH 11/14] nl80211: add RNR for EMA AP

Store the reduced neighbor report (RNR) elements provided by application.
RNR is split into multiple elements, each corresponding to one
multiple BSSID element in the kernel. Both are added to create
EMA (Enhanced multi-BSSID advertisements) beacons.
Application is responsible of creating blocks such that if a
non-transmitted profile is included in multiple BSSID element then
it is excluded from the RNR of the same beacon.
RNR block for each EMA may also include entries common for all beacons
such as co-located APs.
Offsets are not used if EMA is not enabled.

Signed-off-by: Aloka Dixit <quic_alokad@quicinc.com>
---
 include/net/cfg80211.h | 21 ++++++++++++++++++
 net/wireless/nl80211.c | 50 ++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 71 insertions(+)

--- a/include/net/cfg80211.h
+++ b/include/net/cfg80211.h
@@ -1175,6 +1175,25 @@ struct cfg80211_mbssid_elems {
 };
 
 /**
+ * struct cfg80211_rnr_elems - Reduced neighbor report (RNR) elements
+ * @ies: array of extra information element(s) to add into Beacon frames for
+ *	RNR or %NULL.
+ *	This is used only when EMA (Enhanced multi-BSSID advertisements) feature
+ *	is enabled and profile periodicity is more than 1.
+ *
+ * @elem: Array of RNR element(s) to be added into Beacon frames.
+ * @elem.data: Data for RNR elements.
+ * @elem.len: Length of data.
+ */
+struct cfg80211_rnr_elems {
+	u8 cnt;
+	struct {
+		const u8 *data;
+		size_t len;
+	} elem[];
+};
+
+/**
  * struct cfg80211_beacon_data - beacon data
  * @head: head portion of beacon (before TIM IE)
  *	or %NULL if not changed
@@ -1193,6 +1212,7 @@ struct cfg80211_mbssid_elems {
  * @probe_resp_len: length of probe response template (@probe_resp)
  * @probe_resp: probe response template (AP mode only)
  * @mbssid_ies: multiple BSSID elements
+ * @rnr_ies: reduced neighbor report elements
  * @ftm_responder: enable FTM responder functionality; -1 for no change
  *	(which also implies no change in LCI/civic location data)
  * @lci: Measurement Report element content, starting with Measurement Token
@@ -1211,6 +1231,7 @@ struct cfg80211_beacon_data {
 	const u8 *lci;
 	const u8 *civicloc;
 	struct cfg80211_mbssid_elems *mbssid_ies;
+	struct cfg80211_rnr_elems *rnr_ies;
 	s8 ftm_responder;
 
 	size_t head_len, tail_len;
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -809,6 +809,7 @@ static const struct nla_policy nl80211_p
 	[NL80211_ATTR_RADAR_BACKGROUND] = { .type = NLA_FLAG },
 	[NL80211_ATTR_AP_SETTINGS_FLAGS] = { .type = NLA_U32 },
 	[NL80211_ATTR_6G_REG_POWER_MODE] = NLA_POLICY_RANGE(NLA_U8, 0, 2),
+	[NL80211_ATTR_EMA_RNR_ELEMS] = { .type = NLA_NESTED },
 #if LINUX_VERSION_IS_GEQ(5,10,0)
 	[NL80211_ATTR_EHT_CAPABILITY] =
 		NLA_POLICY_RANGE(NLA_BINARY,
@@ -5414,6 +5415,30 @@ nl80211_parse_mbssid_elems(struct wiphy
 	return elems;
 }
 
+static struct cfg80211_rnr_elems *
+nl80211_parse_rnr_elems(struct wiphy *wiphy, struct nlattr *attrs)
+{
+	struct nlattr *nl_elems;
+	struct cfg80211_rnr_elems *elems;
+	int rem_elems;
+	u8 i = 0, num_elems = 0;
+
+	nla_for_each_nested(nl_elems, attrs, rem_elems)
+		num_elems++;
+
+	elems = kzalloc(struct_size(elems, elem, num_elems), GFP_KERNEL);
+	if (!elems)
+		return ERR_PTR(-ENOMEM);
+
+	nla_for_each_nested(nl_elems, attrs, rem_elems) {
+		elems->elem[i].data = nla_data(nl_elems);
+		elems->elem[i].len = nla_len(nl_elems);
+		i++;
+	}
+	elems->cnt = num_elems;
+	return elems;
+}
+
 static int nl80211_parse_beacon(struct cfg80211_registered_device *rdev,
 				struct nlattr *attrs[],
 				struct cfg80211_beacon_data *bcn)
@@ -5503,6 +5528,20 @@ static int nl80211_parse_beacon(struct c
 			return PTR_ERR(mbssid);
 
 		bcn->mbssid_ies = mbssid;
+
+		if (bcn->mbssid_ies && attrs[NL80211_ATTR_EMA_RNR_ELEMS]) {
+			struct cfg80211_rnr_elems *rnr =
+				nl80211_parse_rnr_elems(&rdev->wiphy,
+						attrs[NL80211_ATTR_EMA_RNR_ELEMS]);
+
+			if (IS_ERR(rnr))
+				return PTR_ERR(rnr);
+
+			bcn->rnr_ies = rnr;
+			if (bcn->rnr_ies &&
+			    bcn->rnr_ies->cnt < bcn->mbssid_ies->cnt)
+				return -EINVAL;
+		}
 	}
 
 	return 0;
@@ -6006,6 +6045,11 @@ static int nl80211_start_ap(struct sk_bu
 							0);
 		if (err)
 			goto out;
+
+		if (!params->mbssid_config.ema && params->beacon.rnr_ies) {
+			err = -EINVAL;
+			goto out;
+		}
 	}
 
 	nl80211_calculate_ap_params(params);
@@ -6037,6 +6081,7 @@ out:
 	    params->mbssid_config.tx_wdev->netdev &&
 	    params->mbssid_config.tx_wdev->netdev != dev)
 		dev_put(params->mbssid_config.tx_wdev->netdev);
+	kfree(params->beacon.rnr_ies);
 	kfree(params);
 
 	return err;
@@ -6070,6 +6115,7 @@ static int nl80211_set_beacon(struct sk_
 
 out:
 	kfree(params.mbssid_ies);
+	kfree(params.rnr_ies);
 	return err;
 }
 
@@ -9947,6 +9993,8 @@ skip_beacons:
 free:
 	kfree(params.beacon_after.mbssid_ies);
 	kfree(params.beacon_csa.mbssid_ies);
+	kfree(params.beacon_after.rnr_ies);
+	kfree(params.beacon_csa.rnr_ies);
 	kfree(csa_attrs);
 	return err;
 }
@@ -15494,6 +15542,8 @@ static int nl80211_color_change(struct s
 out:
 	kfree(params.beacon_next.mbssid_ies);
 	kfree(params.beacon_color_change.mbssid_ies);
+	kfree(params.beacon_next.rnr_ies);
+	kfree(params.beacon_color_change.rnr_ies);
 	kfree(tb);
 	return err;
 }
